src/channel-qvm.lisp

```diff
@@ -0,0 +1,204 @@
+  ;; Initializes an instance of a CHANNEL-QVM"
+  (declare (ignore args))
+  (setf
+   ;; Initialize the trial-amplitudes to an empty array of the correct
+   ;; size
+   (%trial-amplitudes qvm) (make-lisp-cflonum-vector (expt 2 (number-of-qubits qvm)))
+   ;; Save a pointer to the originally provided memory
+   (slot-value qvm 'original-amplitude-pointer) (amplitudes qvm)))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+(defun requires-swapping-p (qvm)
+  "Does the  QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  ;; If any noise rules are matched by theinstruction INSTR, apply the
+  ;; the kraus operators for that noise rule after applying the gate
+  ;; for this instruction.
+  (let* ((gate (pull-teeth-to-get-a-gate instr))
+         (params (mapcar (lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-matched-rule (find-matching-rule noise-rules instr ':before))
+         (append-matched-rule (find-matching-rule noise-rules instr ':after)))
+    (when prepend-matched-rule
+      (apply-all-kraus-maps qvm instr (operation-elements prepend-matched-rule)))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-matched-rule
+      (apply-all-kraus-maps qvm instr (operation-elements append-matched-rule)))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun rule-matches-instr-p (rule instr position)
+  "Check if rule is matched by instruction data and the position of the match request."
+  (let* ((noise-predicate (predicate-function (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (eq position noise-position))))
+
+
+(defun find-matching-rule (rules instr position)
+  "Return the first rule that matches INSTR/POSITION in the list RULES."
+  (find-if (lambda (rule) (rule-matches-instr-p rule instr position)) rules))
```

Comment:
Is a list the most appropriate data structure for the rule set? This query is going to be O(n) where n is the size of the rules. Probably OK for this PR but we might need to revisit this in the future.

Reply:
I was also thinking about this, but I agree that for at least this PR a list should be okay. For most use cases that I can think of, the list of noise rules will be small compared to the length of the program.